crates/astria-sequencer/src/checked_actions/init_bridge_account.rs

```diff
@@ -0,0 +1,335 @@
+use astria_core::{
+    primitive::v1::{
+        asset::IbcPrefixed,
+        Address,
+        ADDRESS_LEN,
+    },
+    protocol::transaction::v1::action::InitBridgeAccount,
+};
+use astria_eyre::eyre::{
+    ensure,
+    Result,
+    WrapErr as _,
+};
+use cnidarium::{
+    StateRead,
+    StateWrite,
+};
+use tracing::{
+    instrument,
+    Level,
+};
+
+use super::{
+    AssetTransfer,
+    TransactionSignerAddressBytes,
+};
+use crate::{
+    address::StateReadExt as _,
+    bridge::{
+        StateReadExt as _,
+        StateWriteExt as _,
+    },
+};
+
+#[derive(Debug)]
+pub(crate) struct CheckedInitBridgeAccount {
+    action: InitBridgeAccount,
+    tx_signer: TransactionSignerAddressBytes,
+}
+
+impl CheckedInitBridgeAccount {
+    #[instrument(skip_all, err(level = Level::DEBUG))]
+    pub(super) async fn new<S: StateRead>(
+        action: InitBridgeAccount,
+        tx_signer: [u8; ADDRESS_LEN],
+        state: S,
+    ) -> Result<Self> {
+        // Run immutable checks for base prefix.
+        if let Some(sudo_address) = &action.sudo_address {
+            state
+                .ensure_base_prefix(sudo_address)
+                .await
+                .wrap_err("sudo address has an unsupported prefix")?;
+        }
+        if let Some(withdrawer_address) = &action.withdrawer_address {
+            state
+                .ensure_base_prefix(withdrawer_address)
+                .await
+                .wrap_err("withdrawer address has an unsupported prefix")?;
+        }
+
+        let checked_action = Self {
+            action,
+            tx_signer: tx_signer.into(),
+        };
+        checked_action.run_mutable_checks(state).await?;
+
+        Ok(checked_action)
+    }
+
+    #[instrument(skip_all, err(level = Level::DEBUG))]
+    pub(super) async fn run_mutable_checks<S: StateRead>(&self, state: S) -> Result<()> {
+        // Ensure the tx signer account is not a bridge account.
+        //
+        // This prevents the address from being registered as a bridge account if it's been
+        // previously initialized as a bridge account.
+        //
+        // However, there is no prevention of initializing an account as a bridge account that's
+        // already been used as a normal EOA.
+        //
+        // The implication is that the account might already have a balance, nonce, etc. before
+        // being converted into a bridge account.
+        //
+        // After the account becomes a bridge account, it can no longer receive funds via
+        // `Transfer`, only via `BridgeLock` or `BridgeTransfer`.
```

Comment:
bridge accounts should still be able to receive `Transfer`s, in case balance needs to be topped up to pay fees for example

Reply:
Removed in [73ec272](https://github.com/astriaorg/astria/pull/2142/commits/73ec27210915e8a170768ff722a20d45297b4f81).